napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,167 @@
+        self._lock_layers_to_task = RLock()
+
+    def slice_layers_async(
+        self, layers: Iterable[Layer], dims: Dims
+    ) -> Future[dict]:
+        """This should only be called from the main thread.
+
+        Creates a new task and adds it to the _layers_to_task dict. Cancels
+        all tasks currently pending for that layer tuple.
+
+        Submitting multiple layers at one generates multiple requests (stored in a dict),
+        but only ONE task.
+
+        If multiple layers are sliced, any task that contains only one of those
+        layers can safely be cancelled. If a single layer is sliced, it will
+        wait for any existing tasks that include that layer AND another layer,
+        In other words, it will only cancel if the new task will replace the
+        slices of all the layers in the pending task.
+        """
+        # Cancel any tasks that are slicing a subset of the layers
+        # being sliced now. This allows us to slice arbitrary sets of
+        # layers with some sensible and not too complex cancellation
+        # policy.
+        if existing_task := self._find_existing_task(layers):
+            logger.debug('Cancelling task for %s', layers)
+            existing_task.cancel()
+
+        # Not all layer types will initially be asynchronously sliceable.
+        # The following logic gives us a way to handle those in the short
+        # term as we develop, and also in the long term if there are cases
+        # when we want to perform sync slicing anyway.
+        requests = {}
+        for layer in layers:
+            if layer._is_async():
+                requests[layer] = layer._make_slice_request(dims)
+            else:
+                layer._slice_dims(dims.point, dims.ndisplay, dims.order)
+        # create task for slicing of each request/layer
+        task = self._executor.submit(self._slice_layers, requests)
+
+        # store task for cancellation logic
+        # this is purposefully done before adding the done callback to ensure
+        # that the task is added before the done callback can be executed
+        with self._lock_layers_to_task:
+            self._layers_to_task[tuple(requests.keys())] = task
+
+        task.add_done_callback(self._on_slice_done)
+
+        return task
+
+    def shutdown(self) -> None:
+        """Should be called from the main thread when this is no longer needed."""
+        self._executor.shutdown()
+
+    def _slice_layers(self, requests: Dict) -> Dict:
+        """
+        Iterates through a dictionary of request objects and call the slice
+        on each individual layer. Can be called from the main or slicing thread.
+
+        Attributes
+        ----------
+        requests: dict[Layer, SliceRequest]
+            Dictionary of request objects to be used for constructing the slice
+
+        Returns
+        -------
+        Dict[Layer, SliceResponse] which contains the results of the slice
```

Comment:
```suggestion
        dict[Layer, SliceResponse]: which contains the results of the slice
```
This is my opinion. If I correctly understand dict here is not a normal word, but name of a specific python structure. 

also after type specifier there should be `:`